### PR TITLE
Viewer creation

### LINKED ISSUE #31: Implement "new 1D hist", "new 2D scat", "new 2D image" commands
Those commands are doing nothing for now

https://github.com/QuantStack/glue-lab/blob/main/src/controlPanel/plugin.ts#L43-L65

### PR DESCRIPTION
**Pull request by martinRenou**
Would fix #31 

- Implement "new viewer" commands 
- Fix tabs data updates
- Drag and drop of datasets

[Screencast from 2023-05-22 15-28-11.webm](https://github.com/QuantStack/glue-lab/assets/21197331/014e1ebe-e22c-43e6-8c85-b32b2a00086e)

Remaining TODO: 
- [x] Table + 3d scatter viewers
- [x] drag and drop datasets
- [x] background title "Drag Data To Plot"

### CODE REVIEW

File: src/viewPanel/sessionWidget.ts [hunk 1]
```diff
@@ -150,23 +150,29 @@ export class SessionWidget extends BoxPanel {
...
-          context: this._context,
-          notebookTracker: this._notebookTracker,
-          dataLoaded: this._dataLoaded
-        });
-        const tabWidget = new TabView({ model });
-
-        this._tabPanel.addTab(tabWidget, idx + 1);
+        if (tabName in this._tabsModels) {
+          this._tabsModels[tabName].tabData = tabData;
+        } else {
```
Comment (martinRenou) [inline src/viewPanel/sessionWidget.ts:155]: @trungleduc @hbcarlos I suppose we'll need to be granular on the changes here?

We don't want to re-render all viewers (plots) every time there is a change in the tabs data. Instead, we want to:
- create new viewers if there are new ones (what my `if` condition is doing here)
- delete old viewers if they are not part of the shared model anymore
- update viewers states if they need updating

How do you see the implementation for this? Should we do it all manual? I suppose the YJS event should contain everything about what actually happens in the tabs change?

Comment (trungleduc) [reply src/viewPanel/sessionWidget.ts:155]: > I suppose the YJS event should contain everything about what actually happens in the tabs change?

I think it should be if the tab key is deep-observed

Comment (hbcarlos) [reply src/viewPanel/sessionWidget.ts:155]: Agree. We should not re-render everything all the time.

> I suppose the YJS event should contain everything about what actually happens in the tabs change?

Yes. Each tab should be a map, and then we can use deepobserve to listen for changes on each tab.

File: src/controlPanel/data/datasetsWidget.tsx [hunk 2]
```diff
@@ -89,12 +94,26 @@ export class DatasetsWidget extends ReactWidget {
         : ''
     }`;
     return (
-      <li id={id} className={className} onClick={this._onClick.bind(this)}>
+      <li
+        id={id}
```
Comment (trungleduc) [inline src/controlPanel/data/datasetsWidget.tsx:98]: It's from the previous code, but why do we need an id for this element?

File: src/viewPanel/tabView.ts [hunk 3]
```diff
@@ -25,12 +38,17 @@ export class TabView extends Widget {
...
-    this._model.ready.connect(() => {
-      this._initGridItems();
+    this._ready = new Signal<this, null>(this);
+    this._ready.connect(() => {
+      this._updateViewers();
+    });
+
+    this._context.sessionContext.ready.then(() => {
+      this._ready.emit(null);
     });
```
Comment (trungleduc) [inline src/viewPanel/tabView.ts:48]: `this._ready` is private without a getter and is not used elsewhere. 

```suggestion
    this._context.sessionContext.ready.then(() => {
      this._updateViewers();
    });
```

File: src/controlPanel/data/datasetsWidget.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinRenou) [reply src/controlPanel/data/datasetsWidget.tsx:98]: Yes, because switching to another session would re-render the control panel.

**Review by trungleduc — APPROVED** [review]

**Pull request merged**